Misleading exception during async getObject

### Describe the bug

I was very confused when I saw this error message, as it implies there was an issue with uploading. This error was thrown as part of an asynchronous getObject request,

```
S3AsyncClient#getObject(ObjectRequest,AsyncResponseTransformer<GetObjectResponse,ReturnT>)
```

Note, "written" and "upload" in the exception message.

```
Caused by: software.amazon.awssdk.core.exception.RetryableException: Data read has a different checksum than expected. Was 0x14173258040ed24cc9dd698d1f43d8cf, but expected 0x001c00001c0000000607000050415231. Common causes: (1) You modified a request ByteBuffer before it could be written to the service. Please ensure your data source does not modify the  byte buffers after you pass them to the SDK. (2) The data was corrupted between the client and service. Note: Despite this error, the upload still completed and was persisted in S3.
                at software.amazon.awssdk.core.exception.RetryableException$BuilderImpl.build(RetryableException.java:99)
                at software.amazon.awssdk.core.exception.RetryableException.create(RetryableException.java:35)
                at software.amazon.awssdk.services.s3.internal.checksums.S3ChecksumValidatingPublisher$ChecksumValidatingSubscriber.onComplete(S3ChecksumValidatingPublisher.java:137)
                at software.amazon.awssdk.core.internal.metrics.BytesReadTrackingPublisher$BytesReadTracker.onComplete(BytesReadTrackingPublisher.java:74)
                at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler$DataCountingPublisher$1.onComplete(ResponseHandler.java:519)
                at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.runAndLogError(ResponseHandler.java:254)
                at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.access$600(ResponseHandler.java:77)
                at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler$PublisherAdapter$1.onComplete(ResponseHandler.java:375)
```


### Regression Issue

- [ ] Select this option if this issue appears to be a regression.

### Expected Behavior

More appropriate error message. (Probably makes sense to have the same error message as the sync case in `.S3ChecksumValidatingInputStream#validateAndThrow`)

### Current Behavior

See description.

### Reproduction Steps

(Not sure how to reproduce invalid checksums on demand.)

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### AWS Java SDK version used

2.29.52

### JDK version used

17.0.16

### Operating System and version

Red Hat 9

--- a/.changes/next-release/bugfix-AWS-5dd0c1d.json
+++ b/.changes/next-release/bugfix-AWS-5dd0c1d.json
@@ -1,0 +1,6 @@
+{
+    "type": "bugfix",
+    "category": "Amazon S3",
+    "contributor": "",
+    "description": "Fixed misleading checksum mismatch error message for S3 GetObject that incorrectly referenced uploading. See [#6324](https://github.com/aws/aws-sdk-java-v2/issues/6324)."
+}

--- a/services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/checksums/ChecksumConstant.java
+++ b/services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/checksums/ChecksumConstant.java
@@ -55,6 +55,13 @@ public final class ChecksumConstant {
      */
     public static final int S3_MD5_CHECKSUM_LENGTH = 16;
 
+    public static final String CHECKSUM_MISMATCH_ERROR_MESSAGE_TEMPLATE =
+        "Data read has a different checksum than expected. Was 0x%s, "
+        + "but expected 0x%s. Common causes: (1) The data was corrupted during transfer between S3 and the client. "
+        + "(2) A custom ExecutionInterceptor modified the HTTP response content before checksum validation. "
+        + "Note: The response data may have already been delivered to the configured response transformer "
+        + "and should not be considered reliable.";
+
     private ChecksumConstant() {
     }
 }

--- a/services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/checksums/S3ChecksumValidatingInputStream.java
+++ b/services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/checksums/S3ChecksumValidatingInputStream.java
@@ -15,6 +15,8 @@
 
 package software.amazon.awssdk.services.s3.internal.checksums;
 
+import static software.amazon.awssdk.services.s3.internal.checksums.ChecksumConstant.CHECKSUM_MISMATCH_ERROR_MESSAGE_TEMPLATE;
+
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.Arrays;
@@ -182,9 +184,8 @@ public class S3ChecksumValidatingInputStream extends InputStream implements Abor
 
         if (!Arrays.equals(computedChecksum, streamChecksum)) {
             throw RetryableException.create(
-                String.format("Data read has a different checksum than expected. Was 0x%s, but expected 0x%s. " +
-                              "This commonly means that the data was corrupted between the client and " +
-                              "service.", BinaryUtils.toHex(computedChecksum), BinaryUtils.toHex(streamChecksum)));
+                String.format(CHECKSUM_MISMATCH_ERROR_MESSAGE_TEMPLATE, BinaryUtils.toHex(computedChecksum),
+                              BinaryUtils.toHex(streamChecksum)));
         }
     }
 

--- a/services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/checksums/S3ChecksumValidatingPublisher.java
+++ b/services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/checksums/S3ChecksumValidatingPublisher.java
@@ -16,6 +16,7 @@
 package software.amazon.awssdk.services.s3.internal.checksums;
 
 import static java.lang.Math.toIntExact;
+import static software.amazon.awssdk.services.s3.internal.checksums.ChecksumConstant.CHECKSUM_MISMATCH_ERROR_MESSAGE_TEMPLATE;
 
 import java.nio.ByteBuffer;
 import java.util.Arrays;
@@ -135,12 +136,7 @@ public final class S3ChecksumValidatingPublisher implements SdkPublisher<ByteBuf
                 byte[] computedChecksum = sdkChecksum.getChecksumBytes();
                 if (!Arrays.equals(computedChecksum, streamChecksum)) {
                     onError(RetryableException.create(
-                        String.format("Data read has a different checksum than expected. Was 0x%s, but expected 0x%s. "
-                                      + "Common causes: (1) You modified a request ByteBuffer before it could be "
-                                      + "written to the service. Please ensure your data source does not modify the "
-                                      + " byte buffers after you pass them to the SDK. (2) The data was corrupted between the "
-                                      + "client and service. Note: Despite this error, the upload still completed and was "
-                                      + "persisted in S3.",
+                        String.format(CHECKSUM_MISMATCH_ERROR_MESSAGE_TEMPLATE,
                                       BinaryUtils.toHex(computedChecksum), BinaryUtils.toHex(streamChecksum))));
                     return; // Return after onError and not call onComplete below
                 }

--- a/services/s3/src/test/java/software/amazon/awssdk/services/s3/checksums/S3ChecksumValidatingInputStreamTest.java
+++ b/services/s3/src/test/java/software/amazon/awssdk/services/s3/checksums/S3ChecksumValidatingInputStreamTest.java
@@ -15,6 +15,7 @@
 
 package software.amazon.awssdk.services.s3.checksums;
 
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.Assertions.assertArrayEquals;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.fail;
@@ -107,7 +108,7 @@ public class S3ChecksumValidatingInputStreamTest {
                 IoUtils.toByteArray(validatingInputStream);
                 fail("Corruption at byte " + i + " was not detected.");
             } catch (SdkClientException e) {
-                // Expected
+                assertThat(e.getMessage()).contains("Data read has a different checksum than expected");
             }
         }
     }

--- a/services/s3/src/test/java/software/amazon/awssdk/services/s3/checksums/S3ChecksumValidatingPublisherTest.java
+++ b/services/s3/src/test/java/software/amazon/awssdk/services/s3/checksums/S3ChecksumValidatingPublisherTest.java
@@ -15,6 +15,7 @@
 
 package software.amazon.awssdk.services.s3.checksums;
 
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.Assertions.assertArrayEquals;
 import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertTrue;
@@ -91,6 +92,7 @@ public class S3ChecksumValidatingPublisherTest {
 
     assertFalse(s.hasCompleted());
     assertTrue(s.isOnErrorCalled());
+    assertThat(s.getError().getMessage()).contains("Data read has a different checksum than expected");
   }
 
   @Test
@@ -195,6 +197,7 @@ public class S3ChecksumValidatingPublisherTest {
     final List<ByteBuffer> received;
     boolean completed;
     boolean onErrorCalled;
+    Throwable error;
 
     TestSubscriber() {
       this.received = new ArrayList<>();
@@ -216,6 +219,7 @@ public class S3ChecksumValidatingPublisherTest {
     @Override
     public void onError(Throwable t) {
       onErrorCalled = true;
+      error = t;
     }
 
     @Override
@@ -241,6 +245,10 @@ public class S3ChecksumValidatingPublisherTest {
 
     public boolean isOnErrorCalled() {
       return onErrorCalled;
+    }
+
+    public Throwable getError() {
+      return error;
     }
   }
 

